### PR TITLE
DivisionRing and Quaternion made lawful.

### PR DESCRIPTION
**Pull request by denisrosset**
Added a DivisionRing typeclass.

Converted Quaternion to DivisionRing; split the Quaternion instances into one requiring only a signed Field scalar, and another richer that include roots but requires a NRoot/Trig scalar.

Added tests using the Quaternion[Rational] type.

Note: I changed some `IsReal` implicits into `Order` / `Signed` to be more relaxed in the requirements. This change will be complete in a future PR.

### CODE REVIEW

**Comment by denisrosset** [comment]
Note: the codecov/patch failure is related to the fromDouble code duplication, which will come later from algebra `Ring.defaultFromDouble`.

File: core/shared/src/main/scala/spire/algebra/DivisionRing.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+  def fromDouble(a: Double): A = DivisionRing.fromDouble[A](a)(self, self)
+}
+
+
+object DivisionRing {
+
+  @inline final def apply[A](implicit f: DivisionRing[A]): DivisionRing[A] = f
+
+  // TODO: use instead algebra Ring.fromDouble, if PR algebra/#190 is accepted
```
Comment (tixxit) [inline core/shared/src/main/scala/spire/algebra/DivisionRing.scala]: This was merged! Can we delete this and use the one from here now? https://github.com/typelevel/algebra/pull/190

File: core/shared/src/main/scala/spire/math/Quaternion.scala [hunk 2]
```diff
@@ -50,33 +44,47 @@ private[math] trait QuaternionAlgebra[A]
...
-    _gcd(a, b)
-  }
-   */
+
+  def timesl(a: A, q: Quaternion[A]): Quaternion[A] = q * a
+  def dot(x: Quaternion[A], y: Quaternion[A]): A = x.dot(y)
+
+}
+
+private[math] trait QuaternionOverRichField[A] extends QuaternionOverField[A]
```
Comment (tixxit) [inline core/shared/src/main/scala/spire/math/Quaternion.scala]: Nice split 👍

File: core/shared/src/main/scala/spire/algebra/DivisionRing.scala (hunk 1 above)
Comment (denisrosset) [inline core/shared/src/main/scala/spire/algebra/DivisionRing.scala]: typelevel/algebra#190 was merged but not published. Can Spire depend on snapshot versions of algebra?

Comment (denisrosset) [inline core/shared/src/main/scala/spire/algebra/DivisionRing.scala]: Let's file an issue to keep track of this todo, and correct it later.

**Comment by tixxit** [comment]
👍

**Comment by denisrosset** [comment]
Thanks! I added an issue #622 to remind us the future code removal.

**Pull request merged**